packages/reason-css-parser/ppx/EmitType.re

```diff
@@ -0,0 +1,301 @@
+    open String;
+    let length = length(str);
+    length >= 2 && sub(str, length - 2, 2) == "()";
+  };
+  let function_value_name = function_name => "function-" ++ function_name;
+  let property_value_name = property_name =>
+    is_function(property_name)
+      ? function_value_name(property_name) : "property-" ++ property_name;
+
+  let value_of_delimiter =
+    fun
+    | "+" => "cross"
+    | "-" => "dash"
+    | "*" => "asterisk"
+    | "/" => "bar"
+    | "@" => "at"
+    | "," => "comma"
+    | ";" => ""
+    | ":" => "doubledot"
+    | "." => "dot"
+    | "(" => "openparen"
+    | ")" => "closeparen"
+    | "[" => "openbracket"
+    | "]" => "closebracket"
+    | "{" => "opencurly"
+    | "}" => "closecurly"
+    | "^" => "caret"
+    | "<" => "lessthan"
+    | "=" => "equal"
+    | ">" => "biggerthan"
+    | "|" => "vbar"
+    | "~" => "tilde"
+    | "$" => "dollar"
+    | _ => "unknown";
+
+  let value_name_of_css = str =>
+    String.(
+      {
+        let length = length(str);
+        let str =
+          if (is_function(str)) {
+            let str = sub(str, 0, length - 2);
+            function_value_name(str);
+          } else {
+            str;
+          };
+        kebab_case_to_snake_case(str);
+      }
+    );
+
+  let abstract_type = name => {
+      type_declaration(
+        ~name=txt(name),
+        ~params=[],
+        ~cstrs=[],
+        ~private_=Public,
+        ~manifest=None,
+        ~kind=Ptype_abstract,
+      );
+  };
+  // TODO: multiplier name
+  let rec variant_name = value => {
+    let value_name =
+      switch (value) {
+      | Terminal(Delim(name), _) => value_of_delimiter(name)
+      | Terminal(Keyword(name), _) => kebab_case_to_snake_case(name)
+      | Terminal(Data_type(name), _) => value_name_of_css(name)
+      | Terminal(Property_type(name), _) =>
+        property_value_name(name) |> value_name_of_css
+      | Group(value, _) => variant_name(value)
+      | Combinator(Static, _) => "static"
+      | Combinator(And, _) => "and"
+      | Combinator(Or, _) => "or"
+      | Combinator(Xor, _) => "xor"
+      | Function_call(name, _) => value_name_of_css(name)
+      };
+    value_name |> first_uppercase;
+  };
+  let variant_names = values => {
+    // TODO: not exactly a fast algorithm
+    let values = values |> List.map(variant_name);
+    let occurrences = (values, value) =>
+      values |> List.filter((==)(value)) |> List.length;
+    values
+    |> List.fold_left(
+         (acc, value) => {
+           let total = occurrences(values, value);
+           let current_values = acc |> List.map(((name, _)) => name);
+           let current = occurrences(current_values, value);
+           let value = total == 1 ? (value, None) : (value, Some(current));
+           [value, ...acc];
+         },
+         [],
+       )
+    |> List.map(
+         fun
+         | (value, None) => value
+         | (value, Some(int)) => value ++ "_" ++ string_of_int(int),
+       )
```

Comment:
Why fold_left and later map?

Reply:
I did not check this code, just copied from `EmitPatch`

Reply:
fair, this code is simple, after mapping the value to it respective variant name, it checks if there is more than one occurrence, if so, the map after the fold will add a number after the name, it is used to represent polymorphism in some functions like rgb:

```
and function_rgb = [%value.rec "
    rgb( [ <extended-percentage> ]{3} [ '/' <alpha-value> ]? )
  | rgb( [ <number> ]{3} [ '/' <alpha-value> ]? )
  | rgb( [ <extended-percentage> ]#{3} [ ',' <alpha-value> ]? )
  | rgb( [ <number> ]#{3} [ ',' <alpha-value> ]? )
"]
```

where all of the definitions above would generate `Function_rgb` at `variant_name`, but their parameters have different types.

other case is `calc_sum` that can have two different operators, 
```
and calc_sum = [%value.rec "<calc-product> [ [ '+' | '-' ] <calc-product> ]*"]
```

Those operators would be `Static_0` and `Static_1`